### DISCUSSION TITLE
Role-based access doesn't differentiate between roles

### DISCUSSION THREAD

**Discussion by arsekil** [keystonejs/keystone#8405] [Questions]
Hello team,

I am trying to figure out a way tha Keystone would differentiate between three, custom roles I made as a `select` field, which has a relationship to a list called Role in my schema.

I used the `usecase-roles` example repo to build on it, but I always after setting up correctly for one role, the `hideCreate` and `hideDelete` settings are messed up. Meaning it hides for the `user` role, but when logging in with admin account, it still the same, like it wouldn't recogniye that it's a different role that's begin logged in.

`export const permissions = {`
`     Admin: ({ session }: ListAccessArgs) => {//console.log('Admin',!!session?.data.role?.admin);`
`            return !!session?.data.role?.admin}, //false`
`     Szerkesztő: ({ session }: ListAccessArgs) => {//console.log('Editor',!!session?.data.role?.editor); `
`            return !!session?.data.role?.editor}, //false`
`     Felhasználó: ({ session }: ListAccessArgs) => {//console.log('User',!!session?.data.role?.user); `
`            return !!session?.data.role?.user}, //false`
`};`

I noticed that this setup, copied from the example repo, always defaults to `false`. So, I had to experiment with it, but when I finally thought and looked like it's working correctly, changed accounts and still as if i was logged in with let's say the `Felhasználó` user role.

Here is the link to -> [my repo](https://github.com/andrejmoltok/wesnoth-keystone) <-

I would appreciate it if you could take a look at it and give some ideas ro tips on how to make it work correctly.

Thank you

**Comment by marekryb** [comment]
Hi,

As far as I can tell, the problems you have are:
1) you are not asking for role in the session object:
<pre>
const { withAuth } = createAuth({
  listKey: 'User',
  identityField: 'email',
  sessionData: 'name createdAt <b>role { id name role }</b>', // <--- add role here
  secretField: 'password',
  ...
</pre>
2) what you get in session object is different to what you think:
<pre>
export type Session = {
  itemId: string;
  listKey: string;
  data: {
    name: string;
    <b>createdAt: Date;</b>
    role?: {
      id: string;
      name: string;
      <s>admin: boolean;</s>
      <s>editor: boolean;</s>
      <s>user: boolean;</s>
      <b>role: string;</b>
    };
  };
};
</pre>
3) and because of that your checks in `permissions` should be changed accordingly

**Reply by arsekil** [reply]
Thank you for pointing out some imporant things I missed. I figured I am better of using checkboxes instead of select fields. 

Most importantly, I managed to solve the problem with checkboxes, so all is good now. Everything is up in my repo.
